File: WebClient/web/shim/GL.cs
// The slice of OpenTK's GL class the client actually calls (about 75 functions), forwarded to WebGL2 through wwwroot/gl.js.
// Handles are small integers that index tables in JS. Desktop-GL-only calls the client relies on (separate vertex-attribute format,
// DSA-style ProgramUniform*, program-interface queries, shader storage buffers) are emulated on the JS side.
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.JavaScript;
using OpenTK.Mathematics;

namespace OpenTK.Graphics.OpenGL;

public static unsafe partial class GL {
    private const string M = "gl";

    // ---- JS imports (implemented in gl.js) -----------------------------------------------------------------------------
    [JSImport("init", M)] internal static partial int JsInit(string canvasId);
    [JSImport("canvasWidth", M)] internal static partial int CanvasWidth();
    [JSImport("canvasHeight", M)] internal static partial int CanvasHeight();
    [JSImport("activeTexture", M)] private static partial void JsActiveTexture(int unit);
    [JSImport("attachShader", M)] private static partial void JsAttachShader(int p, int s);
    [JSImport("bindBuffer", M)] private static partial void JsBindBuffer(int target, int buf);
    [JSImport("bindBufferBase", M)] private static partial void JsBindBufferBase(int target, int index, int buf);
    [JSImport("bindSampler", M)] private static partial void JsBindSampler(int unit, int s);
    [JSImport("bindTexture", M)] private static partial void JsBindTexture(int target, int t);
    [JSImport("bindVertexArray", M)] private static partial void JsBindVertexArray(int v);
    [JSImport("bindVertexBuffer", M)] private static partial void JsBindVertexBuffer(int binding, int buf, int offset, int stride);
    [JSImport("blendFunc", M)] private static partial void JsBlendFunc(int s, int d);
    [JSImport("bufferData", M)] private static partial void JsBufferData(int target, int size, int usage);
    [JSImport("bufferSubData", M)] private static partial void JsBufferSubData(int target, int offset, [JSMarshalAs<JSType.MemoryView>] Span<byte> data);
    [JSImport("clear", M)] private static partial void JsClear(int mask);
    [JSImport("clearColor", M)] private static partial void JsClearColor(float r, float g, float b, float a);
    [JSImport("compileShader", M)] private static partial void JsCompileShader(int s);
    [JSImport("createProgram", M)] private static partial int JsCreateProgram();
    [JSImport("createShader", M)] private static partial int JsCreateShader(int type);
    [JSImport("cullFace", M)] private static partial void JsCullFace(int mode);
    [JSImport("deleteBuffer", M)] private static partial void JsDeleteBuffer(int b);
    [JSImport("deleteVertexArray", M)] private static partial void JsDeleteVertexArray(int v);
    [JSImport("deleteSampler", M)] private static partial void JsDeleteSampler(int s);
    [JSImport("deleteShader", M)] private static partial void JsDeleteShader(int s);
    [JSImport("depthFunc", M)] private static partial void JsDepthFunc(int f);
    [JSImport("depthMask", M)] private static partial void JsDepthMask(bool m);
    [JSImport("detachShader", M)] private static partial void JsDetachShader(int p, int s);
    [JSImport("disable", M)] private static partial void JsDisable(int cap);
    [JSImport("enable", M)] private static partial void JsEnable(int cap);
    [JSImport("drawArrays", M)] private static partial void JsDrawArrays(int mode, int first, int count);
    [JSImport("drawElements", M)] private static partial void JsDrawElements(int mode, int count, int type, int offset);
    [JSImport("enableVertexAttribArray", M)] private static partial void JsEnableVertexAttribArray(int loc);
    [JSImport("genBuffer", M)] private static partial int JsGenBuffer();
    [JSImport("genSampler", M)] private static partial int JsGenSampler();
    [JSImport("genTexture", M)] private static partial int JsGenTexture();
    [JSImport("genVertexArray", M)] private static partial int JsGenVertexArray();
    [JSImport("getActiveUniformName", M)] private static partial string JsGetActiveUniformName(int p, int i);
    [JSImport("getActiveUniformSize", M)] private static partial int JsGetActiveUniformSize(int p, int i);
    [JSImport("getActiveUniformType", M)] private static partial int JsGetActiveUniformType(int p, int i);
    [JSImport("getAttribLocation", M)] private static partial int JsGetAttribLocation(int p, string name);
    [JSImport("getProgrami", M)] private static partial int JsGetProgrami(int p, int pname);
    [JSImport("getProgramInfoLog", M)] private static partial string JsGetProgramInfoLog(int p);
    [JSImport("getUniformBlockCount", M)] private static partial int JsGetUniformBlockCount(int p);
    [JSImport("getUniformBlockName", M)] private static partial string JsGetUniformBlockName(int p, int i);
    [JSImport("getShaderi", M)] private static partial int JsGetShaderi(int s, int pname);
    [JSImport("getShaderInfoLog", M)] private static partial string JsGetShaderInfoLog(int s);
    [JSImport("getUniformLocation", M)] private static partial int JsGetUniformLocation(int p, string name);
    [JSImport("linkProgram", M)] private static partial void JsLinkProgram(int p);
    [JSImport("programUniform1i", M)] private static partial void JsProgramUniform1i(int p, int loc, int v);
    [JSImport("programUniform1f", M)] private static partial void JsProgramUniform1f(int p, int loc, float v);
    [JSImport("programUniformNf", M)] private static partial void JsProgramUniformNf(int p, int loc, int n, int count, [JSMarshalAs<JSType.MemoryView>] Span<byte> v);
    [JSImport("programUniformMatrix4f", M)] private static partial void JsProgramUniformMatrix4f(int p, int loc, int count, bool transpose, [JSMarshalAs<JSType.MemoryView>] Span<byte> v);
    [JSImport("samplerParameteri", M)] private static partial void JsSamplerParameteri(int s, int pname, int v);
    [JSImport("shaderSource", M)] private static partial void JsShaderSource(int s, string src);
    [JSImport("texStorage2D", M)] private static partial void JsTexStorage2D(int target, int levels, int ifmt, int w, int h);
    [JSImport("texSubImage2D", M)] private static partial void JsTexSubImage2D(int target, int level, int x, int y, int w, int h, int format, int type, [JSMarshalAs<JSType.MemoryView>] Span<byte> data);
    [JSImport("useProgram", M)] private static partial void JsUseProgram(int p);
    [JSImport("vertexAttribBinding", M)] private static partial void JsVertexAttribBinding(int loc, int binding);
    [JSImport("vertexAttribFormat", M)] private static partial void JsVertexAttribFormat(int loc, int size, int type, bool norm, int rel);
    [JSImport("vertexAttribIFormat", M)] private static partial void JsVertexAttribIFormat(int loc, int size, int type, int rel);
    [JSImport("vertexAttribPointer", M)] private static partial void JsVertexAttribPointer(int loc, int size, int type, bool norm, int stride, int offset);
    [JSImport("vertexBindingDivisor", M)] private static partial void JsVertexBindingDivisor(int binding, int divisor);
    [JSImport("viewport", M)] private static partial void JsViewport(int x, int y, int w, int h);
    [JSImport("createStorageTexture", M)] private static partial int JsCreateStorageTexture(int bytes);
    [JSImport("storageTextureData", M)] private static partial void JsStorageTextureData(int id, [JSMarshalAs<JSType.MemoryView>] Span<int> data);
    [JSImport("bindStorageTexture", M)] private static partial void JsBindStorageTexture(int index, int id);

    // ---- OpenTK-shaped API ---------------------------------------------------------------------------------------------
    public static void ActiveTexture(TextureUnit unit) => JsActiveTexture((int)unit);
    public static void AttachShader(int program, int shader) => JsAttachShader(program, shader);
    public static void BindBuffer(BufferTarget target, int buffer) => JsBindBuffer((int)target, buffer);
    public static void BindBufferBase(BufferTarget target, uint index, int buffer) => JsBindBufferBase((int)target, (int)index, buffer);
    public static void BindSampler(uint unit, int sampler) => JsBindSampler((int)unit, sampler);
    public static void BindTexture(TextureTarget target, int texture) => JsBindTexture((int)target, texture);
    public static void BindVertexArray(int vao) => JsBindVertexArray(vao);
    public static void BindVertexBuffer(uint binding, int buffer, nint offset, int stride) => JsBindVertexBuffer((int)binding, buffer, (int)offset, stride);
    public static void BlendFunc(BlendingFactor src, BlendingFactor dst) => JsBlendFunc((int)src, (int)dst);
    public static void BufferData(BufferTarget target, int size, IntPtr data, BufferUsage usage) => JsBufferData((int)target, size, (int)usage);

    public static void BufferSubData<T>(BufferTarget target, int offset, int size, ReadOnlySpan<T> data) where T : unmanaged {
        var bytes = MemoryMarshal.AsBytes(data);
        JsBufferSubData((int)target, offset, MemoryMarshal.CreateSpan(ref MemoryMarshal.GetReference(bytes), Math.Min(size, bytes.Length)));
    }

    public static void Clear(ClearBufferMask mask) => JsClear((int)mask);
    public static void ClearColor(float r, float g, float b, float a) => JsClearColor(r, g, b, a);
    public static void CompileShader(int shader) => JsCompileShader(shader);
    public static int CreateProgram() => JsCreateProgram();
    public static int CreateShader(ShaderType type) => JsCreateShader((int)type);
    public static void CullFace(TriangleFace face) => JsCullFace((int)face);
    public static void DeleteBuffer(int buffer) => JsDeleteBuffer(buffer);
    public static void DeleteVertexArray(int vao) => JsDeleteVertexArray(vao);
    public static void DeleteSampler(int sampler) => JsDeleteSampler(sampler);
    public static void DeleteShader(int shader) => JsDeleteShader(shader);
    public static void DepthFunc(DepthFunction func) => JsDepthFunc((int)func);
    public static void DepthMask(bool flag) => JsDepthMask(flag);
    public static void DetachShader(int program, int shader) => JsDetachShader(program, shader);
    public static void Disable(EnableCap cap) => JsDisable((int)cap);
    public static void Enable(EnableCap cap) => JsEnable((int)cap);
    public static void DrawArrays(PrimitiveType mode, int first, int count) => JsDrawArrays((int)mode, first, count);
    public static void DrawElements(PrimitiveType mode, int count, DrawElementsType type, nint offset) => JsDrawElements((int)mode, count, (int)type, (int)offset);
    public static void EnableVertexAttribArray(int index) => JsEnableVertexAttribArray(index);
    public static void EnableVertexAttribArray(uint index) => JsEnableVertexAttribArray((int)index);
    public static void GenBuffer(out int buffer) => buffer = JsGenBuffer();
    public static int GenBuffer() => JsGenBuffer();
    public static void GenSampler(out int sampler) => sampler = JsGenSampler();
    public static int GenTexture() => JsGenTexture();
    public static int GenVertexArray() => JsGenVertexArray();

    public static string GetActiveUniform(int program, uint index, int bufSize, out int length, out int size, out UniformType type) {
        var name = JsGetActiveUniformName(program, (int)index);
        length = name.Length;
        size = JsGetActiveUniformSize(program, (int)index);
        type = (UniformType)JsGetActiveUniformType(program, (int)index);
        return name;
    }

    public static int GetAttribLocation(int program, string name) => JsGetAttribLocation(program, name);
    public static void GetProgrami(int program, ProgramProperty pname, out int value) => value = JsGetProgrami(program, (int)pname);
    public static void GetProgramInfoLog(int program, out string info) => info = JsGetProgramInfoLog(program);

    public static void GetProgramInterfacei(int program, ProgramInterface iface, ProgramInterfacePName pname, out int value) {
        var count = JsGetUniformBlockCount(program);
        if (pname == ProgramInterfacePName.ActiveResources) { value = count; return; }
        var max = 0;
        for (var i = 0; i < count; i++) max = Math.Max(max, JsGetUniformBlockName(program, i).Length + 1);
        value = max;
    }

    public static string GetProgramResourceName(int program, ProgramInterface iface, uint index, int bufSize, out int length) {
        var name = JsGetUniformBlockName(program, (int)index);
        length = name.Length;
        return name;
    }

    public static void GetShaderi(int shader, ShaderParameterName pname, out int value) => value = JsGetShaderi(shader, (int)pname);
    public static void GetShaderInfoLog(int shader, out string info) => info = JsGetShaderInfoLog(shader);
    public static int GetUniformLocation(int program, string name) => JsGetUniformLocation(program, name);
    public static void LinkProgram(int program) => JsLinkProgram(program);

    public static void ProgramUniform1i(int program, int location, int v) => JsProgramUniform1i(program, location, v);
    public static void ProgramUniform1f(int program, int location, float v) => JsProgramUniform1f(program, location, v);
    public static void ProgramUniform2f(int program, int location, int count, in Vector2 v) => JsProgramUniformNf(program, location, 2, count, MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref Unsafe.As<Vector2, float>(ref Unsafe.AsRef(in v)), 2)));
    public static void ProgramUniform3f(int program, int location, int count, in Vector3 v) => JsProgramUniformNf(program, location, 3, count, MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref Unsafe.As<Vector3, float>(ref Unsafe.AsRef(in v)), 3)));
    public static void ProgramUniform4f(int program, int location, int count, in Vector4 v) => JsProgramUniformNf(program, location, 4, count, MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref Unsafe.As<Vector4, float>(ref Unsafe.AsRef(in v)), 4)));
    public static void ProgramUniformMatrix4f(int program, int location, int count, bool transpose, in Matrix4 m) => JsProgramUniformMatrix4f(program, location, count, transpose, MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref Unsafe.As<Matrix4, float>(ref Unsafe.AsRef(in m)), 16)));

    public static void SamplerParameterIi(int sampler, SamplerParameterI pname, in int value) => JsSamplerParameteri(sampler, (int)pname, value);
    public static void ShaderSource(int shader, string source) => JsShaderSource(shader, source);
    public static void TexStorage2D(TextureTarget target, int levels, SizedInternalFormat format, int width, int height) => JsTexStorage2D((int)target, levels, (int)format, width, height);

    public static void TexSubImage2D<T>(TextureTarget target, int level, int x, int y, int width, int height, PixelFormat format, PixelType type, ReadOnlySpan<T> pixels) where T : unmanaged {
        var bytes = MemoryMarshal.AsBytes(pixels);
        JsTexSubImage2D((int)target, level, x, y, width, height, (int)format, (int)type, MemoryMarshal.CreateSpan(ref MemoryMarshal.GetReference(bytes), bytes.Length));
    }

    public static void UseProgram(int program) => JsUseProgram(program);
    public static void VertexAttribBinding(int attrib, uint binding) => JsVertexAttribBinding(attrib, (int)binding);
    public static void VertexAttribBinding(uint attrib, uint binding) => JsVertexAttribBinding((int)attrib, (int)binding);
    public static void VertexAttribFormat(int attrib, int size, VertexAttribType type, bool normalized, int relativeOffset) => JsVertexAttribFormat(attrib, size, (int)type, normalized, relativeOffset);
    public static void VertexAttribFormat(uint attrib, int size, VertexAttribType type, bool normalized, uint relativeOffset) => JsVertexAttribFormat((int)attrib, size, (int)type, normalized, (int)relativeOffset);
    public static void VertexAttribIFormat(int attrib, int size, VertexAttribIType type, int relativeOffset) => JsVertexAttribIFormat(attrib, size, (int)type, relativeOffset);
    public static void VertexAttribIFormat(uint attrib, int size, VertexAttribIType type, uint relativeOffset) => JsVertexAttribIFormat((int)attrib, size, (int)type, (int)relativeOffset);
    public static void VertexAttribPointer(int attrib, int size, VertexAttribPointerType type, bool normalized, int stride, nint offset) => JsVertexAttribPointer(attrib, size, (int)type, normalized, stride, (int)offset);
    public static void VertexAttribPointer(uint attrib, int size, VertexAttribPointerType type, bool normalized, int stride, nint offset) => JsVertexAttribPointer((int)attrib, size, (int)type, normalized, stride, (int)offset);
    public static void VertexBindingDivisor(uint binding, uint divisor) => JsVertexBindingDivisor((int)binding, (int)divisor);
    public static void Viewport(int x, int y, int width, int height) => JsViewport(x, y, width, height);

    // ---- storage-buffer emulation (used by StorageBuffer<T>) -------------------------------------------------------------
    public static int CreateStorageTexture(int bytes) => JsCreateStorageTexture(bytes);
    public static void StorageTextureData(int handle, ReadOnlySpan<int> data) => JsStorageTextureData(handle, MemoryMarshal.CreateSpan(ref MemoryMarshal.GetReference(data), data.Length));
    public static void BindStorageTexture(uint index, int handle) => JsBindStorageTexture((int)index, handle);
}

File: WebClient/web/wwwroot/gl.js
// WebGL2 implementation of the GL calls the client makes (see shim/GL.cs). Handles are small integers indexing tables below (0 = null).
// Desktop-GL features WebGL2 lacks are emulated here:
//   * separate attribute format API (glVertexAttribFormat / glBindVertexBuffer / glVertexAttribBinding / glVertexBindingDivisor):
//     recorded per VAO and turned into vertexAttribPointer calls just before each draw;
//   * DSA-style glProgramUniform*: temporarily switch programs;
//   * program-interface queries for uniform blocks;
//   * shader storage buffers: an RGBA32UI data texture bound to a fixed unit, read with texelFetch (see the ported shaders).
let gl = null;
let canvas = null;

const bufs = [null], progs = [null], shaders = [null], texs = [null], samplers = [null], vaos = [null], ulocs = [null], storages = [null];
const reg = (arr, o) => { arr.push(o); return arr.length - 1; };
// .NET's MemoryView (Span marshalled to JS) is not itself a typed array: _unsafe_create_view() gives a zero-copy typed-array view over WASM
// memory that is valid only during the call (fine, everything below uses it immediately); slice() is the copying fallback.
const view_of = (v) => (v._unsafe_create_view ? v._unsafe_create_view() : v.slice());

let curProg = null;          // WebGLProgram currently in use
let curProgId = 0;
let curVao = null;           // my VAO wrapper (or null = default)
const ssboBindings = [];     // index -> storage texture id
const SSBO_W = 2048;         // texels per row - must match tools/port_shaders.py
const SSBO_UNIT_BASE = 14;   // storage texture for binding index i lives on texture unit 14 - i

export function init(canvasId) {
    canvas = document.getElementById(canvasId);
    gl = canvas.getContext('webgl2', { alpha: false, antialias: false, depth: true, stencil: true, premultipliedAlpha: false, powerPreference: 'high-performance', preserveDrawingBuffer: false });
    if (!gl) return 0;
    window.__gl = gl;
    gl.pixelStorei(gl.UNPACK_ALIGNMENT, 1);
    return 1;
}
export const canvasWidth = () => canvas.width;
export const canvasHeight = () => canvas.height;

// ---- simple pass-throughs ---------------------------------------------------------------------------------------------
export const activeTexture = (unit) => gl.activeTexture(unit);
export const attachShader = (p, s) => { if (shaders[s]._src && gl.getShaderParameter(shaders[s], gl.SHADER_TYPE) === gl.VERTEX_SHADER) progs[p]._vs = shaders[s]._src; gl.attachShader(progs[p], shaders[s]); };
export const detachShader = (p, s) => gl.detachShader(progs[p], shaders[s]);
export const bindSampler = (unit, s) => gl.bindSampler(unit, s ? samplers[s] : null);
export const bindTexture = (target, t) => gl.bindTexture(target, t ? texs[t] : null);
export const blendFunc = (s, d) => gl.blendFunc(s, d);
export const clear = (mask) => gl.clear(mask);
export const clearColor = (r, g, b, a) => gl.clearColor(r, g, b, a);
export const cullFace = (mode) => gl.cullFace(mode);
export const depthFunc = (f) => gl.depthFunc(f);
export const depthMask = (m) => gl.depthMask(m);
export const viewport = (x, y, w, h) => gl.viewport(x, y, w, h);
export const genBuffer = () => reg(bufs, gl.createBuffer());
export const genSampler = () => reg(samplers, gl.createSampler());
export const genTexture = () => reg(texs, gl.createTexture());
export const deleteBuffer = (b) => { if (bufs[b]) gl.deleteBuffer(bufs[b]); };
export const deleteSampler = (s) => { if (samplers[s]) gl.deleteSampler(samplers[s]); };
export const deleteShader = (s) => { if (shaders[s]) gl.deleteShader(shaders[s]); };
export const samplerParameteri = (s, pname, v) => gl.samplerParameteri(samplers[s], pname, v);
export const texStorage2D = (target, levels, ifmt, w, h) => gl.texStorage2D(target, levels, ifmt, w, h);
export const texSubImage2D = (target, level, x, y, w, h, format, type, data) => gl.texSubImage2D(target, level, x, y, w, h, format, type, view_of(data));

const CAPS = new Set([gl_const('BLEND'), gl_const('CULL_FACE'), gl_const('DEPTH_TEST'), gl_const('STENCIL_TEST'), gl_const('SCISSOR_TEST'),
    gl_const('POLYGON_OFFSET_FILL'), gl_const('DITHER'), gl_const('RASTERIZER_DISCARD'), gl_const('SAMPLE_ALPHA_TO_COVERAGE'), gl_const('SAMPLE_COVERAGE')]);
function gl_const(name) { return ({ BLEND: 0x0BE2, CULL_FACE: 0x0B44, DEPTH_TEST: 0x0B71, STENCIL_TEST: 0x0B90, SCISSOR_TEST: 0x0C11, POLYGON_OFFSET_FILL: 0x8037, DITHER: 0x0BD0, RASTERIZER_DISCARD: 0x8C89, SAMPLE_ALPHA_TO_COVERAGE: 0x809E, SAMPLE_COVERAGE: 0x80A0 })[name]; }
export const enable = (cap) => { if (CAPS.has(cap)) gl.enable(cap); };     // desktop-only caps (FRAMEBUFFER_SRGB, debug output...) are ignored
export const disable = (cap) => { if (CAPS.has(cap)) gl.disable(cap); };

// ---- buffers ------------------------------------------------------------------------------------------------------------
export const bindBuffer = (target, b) => gl.bindBuffer(target, b ? bufs[b] : null);
export const bufferData = (target, size, usage) => gl.bufferData(target, size, usage);
export const bufferSubData = (target, offset, view) => gl.bufferSubData(target, offset, view_of(view));
export function bindBufferBase(target, index, b) {
    if (target === 0x90D2) return;   // shader storage buffer: emulated by storage textures (StorageBuffer<T>), nothing to bind
    gl.bindBufferBase(target, index, b ? bufs[b] : null);
}

// ---- vertex arrays with the separate attribute-format API -----------------------------------------------------------
function makeVao() {
    const attribs = [], bindings = [];
    for (let i = 0; i < 16; i++) {
        attribs.push({ enabled: false, size: 4, type: 0x1406, norm: false, rel: 0, isInt: false, binding: 0, direct: false });
        bindings.push({ buf: null, offset: 0, stride: 0, divisor: 0 });
    }
    return { native: gl.createVertexArray(), attribs, bindings, dirty: false };
}
export const genVertexArray = () => reg(vaos, makeVao());
export const deleteVertexArray = (id) => { const v = vaos[id]; if (v) { if (curVao === v) { curVao = null; gl.bindVertexArray(null); } gl.deleteVertexArray(v.native); vaos[id] = null; } };
export function bindVertexArray(id) {
    curVao = id ? vaos[id] : null;
    gl.bindVertexArray(curVao ? curVao.native : null);
}
export function enableVertexAttribArray(loc) {
    if (curVao) { curVao.attribs[loc].enabled = true; curVao.dirty = true; } else gl.enableVertexAttribArray(loc);
}
export function vertexAttribFormat(loc, size, type, norm, rel) {
    if (!curVao) return; const a = curVao.attribs[loc]; a.size = size; a.type = type; a.norm = norm; a.rel = rel; a.isInt = false; a.direct = false; curVao.dirty = true;
}
export function vertexAttribIFormat(loc, size, type, rel) {
    if (!curVao) return; const a = curVao.attribs[loc]; a.size = size; a.type = type; a.norm = false; a.rel = rel; a.isInt = true; a.direct = false; curVao.dirty = true;
}
export function vertexAttribBinding(loc, binding) { if (!curVao) return; curVao.attribs[loc].binding = binding; curVao.dirty = true; }
export function bindVertexBuffer(binding, b, offset, stride) {
    if (!curVao) return; const v = curVao.bindings[binding]; v.buf = b ? bufs[b] : null; v.offset = offset; v.stride = stride; curVao.dirty = true;
}
export function vertexBindingDivisor(binding, d) { if (!curVao) return; curVao.bindings[binding].divisor = d; curVao.dirty = true; }
export function vertexAttribPointer(loc, size, type, norm, stride, offset) {   // classic API: uses the current ARRAY_BUFFER binding
    gl.vertexAttribPointer(loc, size, type, norm, stride, offset);
    gl.enableVertexAttribArray(loc);
    if (curVao) curVao.attribs[loc].direct = true;
}
function flushVao() {
    const v = curVao;
    if (!v || !v.dirty) return;
    for (let loc = 0; loc < v.attribs.length; loc++) {
        const a = v.attribs[loc];
        if (a.direct) continue;
        if (!a.enabled) continue;
        const b = v.bindings[a.binding];
        if (!b.buf) continue;
        gl.bindBuffer(gl.ARRAY_BUFFER, b.buf);
        if (a.isInt) gl.vertexAttribIPointer(loc, a.size, a.type, b.stride, b.offset + a.rel);
        else gl.vertexAttribPointer(loc, a.size, a.type, a.norm, b.stride, b.offset + a.rel);
        gl.enableVertexAttribArray(loc);
        gl.vertexAttribDivisor(loc, b.divisor);
    }
    v.dirty = false;
}

// ---- storage-buffer emulation -----------------------------------------------------------------------------------------
export function createStorageTexture(bytes) {
    const texels = Math.ceil(bytes / 16);
    const rows = Math.max(1, Math.ceil(texels / SSBO_W));
    const t = gl.createTexture();
    gl.activeTexture(gl.TEXTURE0 + 15);
    gl.bindTexture(gl.TEXTURE_2D, t);
    gl.texStorage2D(gl.TEXTURE_2D, 1, gl.RGBA32UI, SSBO_W, rows);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
    return reg(storages, { tex: t, rows });
}
export function storageTextureData(id, view) {
    const s = storages[id];
    const src = view_of(view);
    const count = src.length;                               // int32 count
    const texels = Math.ceil(count / 4);
    const rows = Math.min(s.rows, Math.max(1, Math.ceil(texels / SSBO_W)));
    const padded = new Uint32Array(rows * SSBO_W * 4);
    padded.set(new Uint32Array(src.buffer, src.byteOffset, count).subarray(0, Math.min(count, padded.length)));
    gl.activeTexture(gl.TEXTURE0 + 15);
    gl.bindTexture(gl.TEXTURE_2D, s.tex);
    gl.texSubImage2D(gl.TEXTURE_2D, 0, 0, 0, SSBO_W, rows, gl.RGBA_INTEGER, gl.UNSIGNED_INT, padded);
}
export function bindStorageTexture(index, id) { ssboBindings[index] = id; }
const ssboUniformCache = new Map();
function bindStorageTexturesForDraw() {
    for (let i = 0; i < ssboBindings.length; i++) {
        const id = ssboBindings[i];
        if (!id) continue;
        const unit = SSBO_UNIT_BASE - i;
        gl.activeTexture(gl.TEXTURE0 + unit);
        gl.bindTexture(gl.TEXTURE_2D, storages[id].tex);
        gl.bindSampler(unit, null);
        if (curProg) {
            const key = curProgId + ':' + i;
            let loc = ssboUniformCache.get(key);
            if (loc === undefined) { loc = gl.getUniformLocation(curProg, 'SsboTex' + i); ssboUniformCache.set(key, loc); }
            if (loc) gl.uniform1i(loc, unit);
        }
    }
}

// ---- draws --------------------------------------------------------------------------------------------------------------
const debugDraws = new URLSearchParams(location.search).has('gldebug');
function checkDraw(kind, mode, count) {   // ?gldebug=1 : report the first GL error after a draw, once per program
    if (!debugDraws) return;
    const e = gl.getError();
    if (e === 0) return;
    const key = curProgId + ':' + e;
    if (checkDraw.seen === undefined) checkDraw.seen = new Set();
    if (checkDraw.seen.has(key)) return;
    checkDraw.seen.add(key);
    const vs = (progs[curProgId] && progs[curProgId]._vs) || '';
    console.log('[gldebug] ' + kind + ' error 0x' + e.toString(16) + ' program #' + curProgId + ' mode ' + mode + ' count ' + count + ' vs: ' + vs.split(String.fromCharCode(10)).slice(5, 9).join(' | '));
}
export function drawArrays(mode, first, count) { flushVao(); bindStorageTexturesForDraw(); gl.drawArrays(mode, first, count); checkDraw('drawArrays', mode, count); }
export function drawElements(mode, count, type, offset) { flushVao(); bindStorageTexturesForDraw(); gl.drawElements(mode, count, type, offset); checkDraw('drawElements', mode, count); }

// ---- shaders / programs ------------------------------------------------------------------------------------------------
export const createShader = (type) => reg(shaders, gl.createShader(type));
export const createProgram = () => reg(progs, gl.createProgram());
export const shaderSource = (s, src) => { shaders[s]._src = src; gl.shaderSource(shaders[s], src); };
export const compileShader = (s) => gl.compileShader(shaders[s]);
export const getShaderi = (s, pname) => { const v = gl.getShaderParameter(shaders[s], pname); return v === true ? 1 : v === false ? 0 : v; };
export const getShaderInfoLog = (s) => gl.getShaderInfoLog(shaders[s]) || '';
export function linkProgram(p) {
    const prog = progs[p];
    gl.linkProgram(prog);
    if (gl.getProgramParameter(prog, gl.LINK_STATUS)) {
        // uniform block i -> binding i, so BindBufferBase(UniformBuffer, i, buf) reaches block i (desktop uses layout(binding=..))
        const n = gl.getProgramParameter(prog, gl.ACTIVE_UNIFORM_BLOCKS);
        for (let i = 0; i < n; i++) gl.uniformBlockBinding(prog, i, i);
    }
}
export function getProgrami(p, pname) {
    const prog = progs[p];
    if (pname === 0x8B87) {   // ACTIVE_UNIFORM_MAX_LENGTH (not available in WebGL): compute
        const n = gl.getProgramParameter(prog, gl.ACTIVE_UNIFORMS); let max = 0;
        for (let i = 0; i < n; i++) max = Math.max(max, gl.getActiveUniform(prog, i).name.length + 1);
        return max;
    }
    const v = gl.getProgramParameter(prog, pname);
    return v === true ? 1 : v === false ? 0 : v;
}
export const getProgramInfoLog = (p) => gl.getProgramInfoLog(progs[p]) || '';
export const getActiveUniformName = (p, i) => gl.getActiveUniform(progs[p], i).name;
export const getActiveUniformSize = (p, i) => gl.getActiveUniform(progs[p], i).size;
export const getActiveUniformType = (p, i) => gl.getActiveUniform(progs[p], i).type;
export const getAttribLocation = (p, name) => gl.getAttribLocation(progs[p], name);
export const getUniformBlockCount = (p) => gl.getProgramParameter(progs[p], gl.ACTIVE_UNIFORM_BLOCKS);
export const getUniformBlockName = (p, i) => gl.getActiveUniformBlockName(progs[p], i);
const uniformIds = new Map();
export function getUniformLocation(p, name) {
    const key = p + ':' + name;
    let id = uniformIds.get(key);
    if (id !== undefined) return id;
    const loc = gl.getUniformLocation(progs[p], name);
    id = loc ? reg(ulocs, loc) : -1;
    uniformIds.set(key, id);
    return id;
}
export function useProgram(p) { curProgId = p; curProg = p ? progs[p] : null; gl.useProgram(curProg); }
function withProgram(p, fn) {
    const prog = progs[p];
    if (prog === curProg) { fn(); return; }
    gl.useProgram(prog); fn(); gl.useProgram(curProg);
}
export function programUniform1i(p, loc, v) { if (loc < 0) return; withProgram(p, () => gl.uniform1i(ulocs[loc], v)); }
export function programUniform1f(p, loc, v) { if (loc < 0) return; withProgram(p, () => gl.uniform1f(ulocs[loc], v)); }
function floatsOf(view, n) {   // view is a Uint8Array over wasm memory holding n floats; copy so alignment never matters
    const out = new Float32Array(n);
    new Uint8Array(out.buffer).set(view_of(view).subarray(0, n * 4));
    return out;
}
export function programUniformNf(p, loc, n, count, view) {
    if (loc < 0) return;
    const f = floatsOf(view, n * count);
    withProgram(p, () => { const u = ulocs[loc]; if (n === 2) gl.uniform2fv(u, f); else if (n === 3) gl.uniform3fv(u, f); else gl.uniform4fv(u, f); });
}
export function programUniformMatrix4f(p, loc, count, transpose, view) {
    if (loc < 0) return;
    let f = floatsOf(view, 16 * count);
    if (transpose) {   // WebGL wants column-major; transpose by hand
        const t = new Float32Array(16 * count);
        for (let m = 0; m < count; m++) for (let r = 0; r < 4; r++) for (let c = 0; c < 4; c++) t[m * 16 + c * 4 + r] = f[m * 16 + r * 4 + c];
        f = t;
    }
    withProgram(p, () => gl.uniformMatrix4fv(ulocs[loc], false, f));
}
